packages/connect-core/src/async-iterable.spec.ts

```diff
@@ -35,6 +37,389 @@ import { Code } from "./code.js";
 import type { EnvelopedMessage } from "./envelope.js";
 import type { Compression } from "./compression.js";
 
+describe("slowly consuming an async iterable", function () {
+  it("should propagate backpressure to source", async function () {
+    const sourceDelayMs = 0;
+    const consumerDelayMs = 50;
+    let sourceElapsedMs = -1;
+    let consumeElapsedMs = -1;
+
+    async function* source() {
+      const tsStart = Date.now();
+      await new Promise((resolve) => setTimeout(resolve, sourceDelayMs));
+      yield "a";
+      await new Promise((resolve) => setTimeout(resolve, sourceDelayMs));
+      yield "b";
+      await new Promise((resolve) => setTimeout(resolve, sourceDelayMs));
+      yield "c";
+      sourceElapsedMs = Date.now() - tsStart;
+    }
+
+    async function slowConsume(source: AsyncIterable<string>) {
+      const tsStart = Date.now();
+      for await (const chunk of source) {
+        expect(chunk).toBeDefined(); // only to satisfy type checks
+        await new Promise((resolve) => setTimeout(resolve, consumerDelayMs));
+      }
+      consumeElapsedMs = Date.now() - tsStart;
+    }
+
+    await slowConsume(source());
+
+    // We expect the source to wait for the consumer, so both should run for
+    // about the same time (3 x 50ms = ~150ms).
+    // We have to be a bit lenient to account for the time it takes run the
+    // actual code, and variance introduced by setTimeout.
+    const leniency = 50;
+    expect(sourceElapsedMs).toBeGreaterThanOrEqual(consumeElapsedMs - leniency);
+    expect(sourceElapsedMs).toBeLessThanOrEqual(consumeElapsedMs + leniency);
+  });
+});
+
+describe("consuming an async iterable with for-await and throwing an error", function () {
+  it("should leave source dangling", async function () {
+    const sourceLog: string[] = [];
+
+    // eslint-disable-next-line @typescript-eslint/require-await
+    async function* source() {
+      try {
+        sourceLog.push("yield a");
+        yield "a";
+        sourceLog.push("yield b");
+        yield "b";
+      } finally {
+        sourceLog.push("finally");
+      }
+    }
+
+    const consumerLog: string[] = [];
+
+    async function consume(source: AsyncIterable<string>) {
+      for await (const chunk of source) {
+        consumerLog.push("received " + chunk);
+        throw "CONSUMER_ERROR";
+      }
+    }
+
+    try {
+      await consume(source());
+      fail("expected error");
+    } catch (e) {
+      expect(e).toBe("CONSUMER_ERROR");
```

Comment:
I'm not sure if this is worth it or not, but you could maybe call `source.throw(e)` here and test that the `finally` was written?

Reply:
This is just the baseline test, proving that a plain for-await loop leaves the iterator dangling, and that `finally` is never reached.

In `pipe()` / `with error-raising consumer`, we actually call `AsyncIterator.throw`, and we assert that this
- `should not propagate error to source by default, and leave it dangling`
- `should propagate error to source with propagateDownStreamError: true`

So I think we're covered pretty well?